cli/tools/registry/mod.rs

```diff
@@ -93,157 +180,333 @@ impl PreparedPublishPackage {
+  pub fn new(
+    graph_diagnostics_collector: GraphDiagnosticsCollector,
+    module_graph_creator: Arc<ModuleGraphCreator>,
+    source_cache: Arc<ParsedSourceCache>,
+    type_checker: Arc<TypeChecker>,
+    fs: Arc<dyn FileSystem>,
+    cli_options: Arc<CliOptions>,
+    mapped_resolver: Arc<MappedSpecifierResolver>,
+  ) -> Self {
+    let sloppy_imports_resolver = if cli_options.unstable_sloppy_imports() {
+      Some(Arc::new(SloppyImportsResolver::new(fs.clone())))
+    } else {
+      None
+    };
+    Self {
+      graph_diagnostics_collector,
+      module_graph_creator,
+      source_cache,
+      type_checker,
+      cli_options,
+      mapped_resolver,
+      sloppy_imports_resolver,
+    }
+  }
+
+  pub async fn prepare_packages_for_publishing(
+    &self,
+    allow_slow_types: bool,
+    diagnostics_collector: &PublishDiagnosticsCollector,
+    deno_json: ConfigFile,
+  ) -> Result<PreparePackagesData, AnyError> {
+    let members = deno_json.to_workspace_members()?;
+
+    if members.len() > 1 {
+      log::info!("Publishing a workspace...");
+    }
+
+    // create the module graph
+    let graph = self
+      .build_and_check_graph_for_publish(
+        allow_slow_types,
+        diagnostics_collector,
+        &members,
+      )
+      .await?;
+
+    let mut package_by_name = HashMap::with_capacity(members.len());
+    let publish_order_graph =
+      publish_order::build_publish_order_graph(&graph, &members)?;
+
+    let results = members
+      .into_iter()
+      .map(|member| {
+        let graph = graph.clone();
+        async move {
+          let package = self
+            .prepare_publish(
+              &member.package_name,
+              &member.config_file,
+              graph,
+              diagnostics_collector,
+            )
+            .await
+            .with_context(|| {
+              format!("Failed preparing '{}'.", member.package_name)
+            })?;
+          Ok::<_, AnyError>((member.package_name, package))
+        }
+        .boxed()
+      })
+      .collect::<Vec<_>>();
+    let results = deno_core::futures::future::join_all(results).await;
+    for result in results {
+      let (package_name, package) = result?;
+      package_by_name.insert(package_name, package);
+    }
+    Ok(PreparePackagesData {
+      publish_order_graph,
+      package_by_name,
+    })
+  }
+
+  async fn build_and_check_graph_for_publish(
+    &self,
+    allow_slow_types: bool,
+    diagnostics_collector: &PublishDiagnosticsCollector,
+    packages: &[WorkspaceMemberConfig],
+  ) -> Result<Arc<deno_graph::ModuleGraph>, deno_core::anyhow::Error> {
+    let build_fast_check_graph = !allow_slow_types;
+    let graph = self
+      .module_graph_creator
+      .create_and_validate_publish_graph(packages, build_fast_check_graph)
+      .await?;
+
+    // todo(dsherret): move to lint rule
+    self
+      .graph_diagnostics_collector
+      .collect_diagnostics_for_graph(&graph, diagnostics_collector)?;
```

Comment:
Do you want to link to https://github.com/denoland/deno_lint/issues/817 which is a prerequisite?

Reply:
It's not necessarily a prerequisite. We already have a cli only lint rule.